esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,432 @@
+            if (sk==StKindAll)
+                throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid Attribute name in AttributeToReturn(%s)", attributeName);
+            if (sk!=StKindNone)
+            {
+                if (!returnStatisticListSpecified)
+                {
+                    returnStatisticList.set(createBitSet(StatisticFilterMaskSize,true));
+                    returnStatisticListSpecified=true;
+                }
+                returnStatisticList->set(sk, true);
+            }
+            else
+            {
+              WuAttr wa = queryWuAttribute(attributeName, WAMax);
+              if (wa!=WAMax)
+              {
+                  if (returnAttributeSpecified)
+                  {
+                      returnAttributeList.set(createBitSet(AttributeFilterMaskSize,true));
+                      returnAttributeSpecified=true;
+                  }
+                  returnAttributeList->set(wa-WANone,true);
+              }
+              else
+                  throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid Attribute name in AttributeToReturn(%s)", attributeName);
+            }
+        }
+    }
+}
+
+void WUDetailsVisitor::resetScope()
+{
+    currentStatisticScopeType = SSTnone;
+    currentAttribs.clear();
+}
+
+static const char * getScopeIdFromScopeName(const char * scopeName)
+{
+    const char * p = strrchr(scopeName, ':');
+    return p?p+1:scopeName;
+}
+
+WUDetails::WUDetails(IConstWorkUnit *_workunit, const char *_wuid, IEspWUDetailsRequest &req)
+: workunit(_workunit), wuid(_wuid), requestScopeFilter(req.getFilter()),
+nestedfilter(req.getNested()), attribsToReturn(req.getAttributeToReturn()),
+scopeOptions(req.getScopeOptions()), attribOptions(req.getAttributeOptions()) { }
+
+void WUDetails::processRequest(IEspWUDetailsResponse &resp)
+{
+    buildWuScopeFilter();
+
+    WuPropertyTypes pt = PTall;
+    if (attribsToReturn.getOnlyDynamic() || *attribsToReturn.getMeasure())
+        pt = PTstatistics;
+
+    IArrayOf<IEspWUResponseScope> respScopes;
+    WUDetailsVisitor wuDetailsVisitor(attribOptions, attribsToReturn);
+    Owned<IConstWUScopeIterator> iter = &workunit->getScopeIterator(wuScopeFilter);
+    ForEach(*iter)
+    {
+        StatisticScopeType scopeType = iter->getScopeType();
+        wuDetailsVisitor.setScopeType(scopeType);
+        iter->playProperties(pt, wuDetailsVisitor);
+
+        const char * scope = iter->queryScope();
+        Owned<IEspWUResponseScope> respScope = createWUResponseScope("","");
+        if (scopeOptions.getIncludeScope())     respScope->setScope(scope);
+        if (scopeOptions.getIncludeScopeType()) respScope->setScopeType(queryScopeTypeName(scopeType));
+        if (scopeOptions.getIncludeId())        respScope->setId(getScopeIdFromScopeName(scope));
+
+        IArrayOf<IEspWUResponseAttribute> & attribs = wuDetailsVisitor.getCurrentAttribs();
+        if (!attribs.empty())
```

Comment:
What happens if I only want a list of scopes?  @GordonSmith is that something you can see you would need?

Reply:
Yes - if I want to get the entire graph structure prior to its execution for example.

Reply:
Done.